ts/api/base_request.ts

```diff
@@ -54,18 +54,28 @@ export abstract class BaseRequest<ResultT, OptionsT> implements
     RelayRequest<ResultT, OptionsT> {
   private promiseResolver = new PromiseResolver<ResultT>();
   private listenerKeys: number[] = [];
-  private timeouts: number[] = [];
   private disposed = false;
+  private timeoutRacer: TimeoutRacer|null = null;
 
   constructor(
       protected frame: ProviderFrameElement,
       protected channel: SecureChannel,
       public id = generateId()) {}
 
-  dispatch(options: OptionsT, timeoutMs?: number): Promise<ResultT> {
-    this.registerBaseHandlers(timeoutMs);
+  async dispatch(options: OptionsT, timeoutRacer?: TimeoutRacer):
+      Promise<ResultT> {
+    this.timeoutRacer = timeoutRacer || startTimeoutRacer(0);
+    this.registerBaseHandlers();
     this.dispatchInternal(options);
-    return this.getPromise();
+    try {
+      return await this.timeoutRacer.race(this.getPromise());
+    } catch (error) {
+      this.timeoutRacer.handleTimeoutError(error);
```

Comment:
Add a comment here that this method will rethrow the error unless it was a timeout error.